controllers/openstackbaremetalset_controller.go

```diff
@@ -95,17 +95,23 @@ func (r *OpenStackBaremetalSetReconciler) Reconcile(ctx context.Context, req ctr
 		return ctrl.Result{}, err
 	}
 
-	// Always patch the instance status when exiting this function so we can persist any changes.
+	// initialize status if Conditions is nil, but do not reset if it already
+	// exists
+	isNewInstance := instance.Status.Conditions == nil
+	if isNewInstance {
+		instance.Status.Conditions = condition.Conditions{}
+	}
+
+	// Save a copy of the condtions so that we can restore the LastTransitionTime
+	// when a condition's state doesn't change.
+	savedConditions := instance.Status.Conditions.DeepCopy()
+
+	// Always patch the instance status when exiting this function so we can
+	// persist any changes.
 	defer func() {
-		// update the Ready condition based on the sub conditions
-		if instance.Status.Conditions.AllSubConditionIsTrue() {
-			instance.Status.Conditions.MarkTrue(
-				condition.ReadyCondition, condition.ReadyMessage)
-		} else {
-			// something is not ready so reset the Ready condition
-			instance.Status.Conditions.MarkUnknown(
-				condition.ReadyCondition, condition.InitReason, condition.ReadyInitMessage)
-			// and recalculate it based on the state of the rest of the conditions
+		condition.RestoreLastTransitionTimes(
```

Comment:
Not sure if I understand this. If during a subsequent reconcile the condition has same state, we restore the last transaction time? Don't we expect something to change to move the condition state to same as before (like when we scaled-out)?

Reply:
During a reconcile, if the previous condition was already True, we want to avoid to move it from True to Unknown to True again (because that particular reconcile doesn't affect it). What we do is calling a lib-common function that restores the previous transition time in the current data structure if the condition hasn't changed (`HasSameState()`). We only update conditions that changed during the reconcile.

Reply:
I understand what the code is doing.. Are we saying that during a reconcile a condition can't go from true->true (with changes to the resource for that condition)? In that case the would not restoring the old transaction time be incorrect.

Reply:
we prevent to go from True to Unknown to True again, because the lastTransitionTime would be updated w/ no reason

Reply:
That's fine if there are no changes to the resource (related to that condition), but there can be changes to the resource for that condition for it to move from (True->(Unknown)->True) for the transaction time to reflect that.

Reply:
If it's ok with you, as we're trying to spread the pattern across the operators, I would move forward with this. Feel free to follow up, and remove that line if it's problematic with the logic of this operator.

Reply:
If we still want to merge irrespective of the issues raised, feel free to do it.

Reply:
The k8s definition of lastTransitionTime is this
lastTransitionTime is the last time the condition transitioned from one status to another. This should be when the underlying condition changed.  If that is not known, then using the time when the API field changed is acceptable.

Reply:
> My initial need for minor updates was a desire to reset conditions

@dprince That's fine. We had that in openstack-operator[1] for a long time before it was removed recently. My point is, we don't have to save conditions and compare for changed state (like we're doing in this patch to get the lastTransitionTime) does not add much value, as that won't be right all the time. As per definition (i.e Last time the condition transitioned from one status to another), it did transition from `unknown` to whatever the reconciliation set it to.

As I mentioned earlier, feel free to merge as is, if the consensus is to move in this direction. 

[1] https://github.com/openstack-k8s-operators/openstack-operator/blob/dev-preview/controllers/core/openstackcontrolplane_controller.go#L150

Reply:
@rabi I think I see your point and the issue you raised. The `RestoreLastTransitionTime` call is something that it might useful if you move from True to Unknown (because we re-init everytime) to True again: in this case the last transition time would be right from a pure reconciliation point of view, but it can be misleading if you look at it. However, the `observedGeneration` can help on that front.
I still think we should go with this code to keep consistency (I'm saying this because I'm pretty sure that line won't cause functional regressions in the operator itself), but I'm ok to remove it in the baremetal operator in a follow up change and revisit the code in others if required.